pkg/storage/cacher.go

```diff
@@ -59,11 +59,56 @@ type CacherConfig struct {
 	// KeyFunc is used to get a key in the underyling storage for a given object.
 	KeyFunc func(runtime.Object) (string, error)
 
+	// IndexFunc is used for optimizing amount of watchers that needs to process
+	// an incoming event.
+	IndexFunc IndexingFunc
+
 	// NewList is a function that creates new empty object storing a list of
 	// objects of type Type.
 	NewListFunc func() runtime.Object
 }
 
+type indexedWatchers struct {
```

Comment:
I think this code would be significantly improved if you made a type for `map[int]*cacheWatcher` instead of basically implementing it twice.